pkg/receive/multitsdb.go

```diff
@@ -801,3 +856,64 @@ func (u *UnRegisterer) MustRegister(cs ...prometheus.Collector) {
 		}
 	}
 }
+
+// extractTenantsLabels extracts tenant's external labels from hashring configs.
+// If one tenant appears in multiple hashring configs,
+// only the external label set from the first hashring config is applied.
+func (t *MultiTSDB) extractTenantsLabels(tenantID string, initialLset labels.Labels) (labels.Labels, error) {
+	for _, hc := range t.hashringConfigs {
+		for _, tenant := range hc.Tenants {
+			if tenant != tenantID {
+				continue
+			}
+
+			if hc.ExternalLabels != nil {
+				extendedLset, err := extendLabels(initialLset, hc.ExternalLabels, t.logger)
+				if err != nil {
+					return nil, errors.Wrap(err, "failed to extend external labels for tenant "+tenantID)
+				}
+				return extendedLset, nil
+			}
+
+			return initialLset, nil
+		}
+	}
+
+	return initialLset, nil
+}
+
+// extendLabels extends external labels of the initial label set.
+// If an external label shares same name with a label in the initial label set,
+// use the label in the initial label set and inform user about it.
+func extendLabels(labelSet labels.Labels, extend map[string]string, logger log.Logger) (labels.Labels, error) {
+	var extendLabels labels.Labels
+	for name, value := range extend {
+		if !model.LabelName.IsValid(model.LabelName(name)) {
+			return nil, errors.Errorf("unsupported format for label's name: %s", name)
+		}
+		extendLabels = append(extendLabels, labels.Label{Name: name, Value: value})
+	}
+
+	extendedLabelSet := make(labels.Labels, 0, len(labelSet)+len(extendLabels))
+	for len(labelSet) > 0 && len(extendLabels) > 0 {
```

Comment:
Will `extendLabels` always be sorted? Otherwise this for loop will not work as expected.

Reply:
I explicitly sorted both the initial label set and extend label set in the implementation.

I also unsorted some external label sets in my test cases. But I'm not sure the label sets in my test cases are sophisticated enough to test all the edge cases of this for loop.